scripts/geninstr.ts

```diff
@@ -482,6 +576,85 @@ declare function value(id: Resource, num?: number): Value;
 declare function value(id: Frame, num?: number): Value;
 declare function value(id: Color, num?: number): Value;
 declare function value(id: Extra, num?: number): Value;
+
+type FromLink = { from: string[] };
+type ToLink = { to: string[] };
+declare function from(...links: string[]): FromLink;
+declare function to(...links: string[]): ToLink;
+
+type LinkValue = Value | AnyId | number;
+type FromLinkArg = FromLink | LinkValue | FromLink[];
+type ToLinkArg = ToLink | LinkValue | (ToLink | LinkValue)[];
+type FromToLinkArg = FromLinkArg | ToLinkArg | (FromLink | ToLink)[];
+
+// foo(v1: Value, v2: Value) -> [FromToLinkArg?, FromToLinkArg?]
+type BehaviorFromToLinkArgs<T> = Partial<{ [K in keyof T]: FromToLinkArg }> | Record<number | string, FromToLinkArg>;
+
+// https://stackoverflow.com/a/52490977
+type Tuple<T, N extends number> = N extends N ? number extends N ? T[] : _TupleOf<T, N, []> : never;
+type _TupleOf<T, N extends number, R extends unknown[]> = R['length'] extends N ? R : _TupleOf<T, N, [T, ...R]>;
+
+declare class Blueprint {
+    type: 'blueprint';
+}
+declare class Component {
+    type: 'component';
+    size: string;
+}
+declare class InternalComponent extends Component {
+    size: 'Internal';
+}
+declare class SmallComponent extends Component {
+    size: 'Small';
+}
+declare class MediumComponent extends Component {
+    size: 'Medium';
+}
+declare class LargeComponent extends Component {
+    size: 'Large';
+}
+
+type InternalComponentArg = InternalComponent;
+type SmallComponentArg = SmallComponent | InternalComponent;
+type MediumComponentArg = MediumComponent | SmallComponent | InternalComponent;
+type LargeComponentArg = LargeComponent | MediumComponent | SmallComponent | InternalComponent;
+
+type BlueprintArgs<I extends number, S extends number, M extends number, L extends number> = {
+    name?: string,
+    power?: boolean,
+    logistics?: {
```

Comment:
Not 100% sure about having these in a nested logistics block in the BlueprintArgs. I think the game blueprint JSON influenced my decision here, and these properties might as well all be top-level. What do you think?

Reply:
Either one seems fine with me. The do correlate with the logistics menu in the game, so having them nested makes sense.

Reply:
I removed the nesting